faster_whisper/transcribe.py

```diff
@@ -1121,51 +1810,414 @@ def find_alignment(
+            if language_probability > language_threshold:
+                num_confident_segments_per_language[language] += 1
+
+                # Add language and probability to the list of languages when it is confident
+                confident_language_probabilities[language].append(language_probability)
+
+                # return the language when sufficient number of confident segments is achieved
+                if (
+                    num_confident_segments_per_language[language]
+                    >= num_detection_segments
+                ):
+                    # Considering the average probability of only confident segments
+                    mean = sum(confident_language_probabilities[language]) / len(
+                        confident_language_probabilities[language]
+                    )
+                    return {
+                        "language_code": language,
+                        "language_confidence": mean,
+                    }
+
+        # if we are unable to get sufficient number of confident predictions,
+        # return the most frequently detected language.
+        # if there is a tie, return the one with maximum average probability.
+        counter = Counter(detected_languages)
+
+        # Define the key function to select frequent language with attached probabilities
+        def key_func(language):
+            # Calculate the frequency of the language
+            frequency = counter[language]
+
+            # Calculate the average probability of the language
+            prob_avg = sum(all_language_probabilities[language]) / len(
+                all_language_probabilities[language]
+            )
+
+            return (frequency, prob_avg)
+
+        max_language = None
+
+        if detected_languages:
+            # Use the key function to find the language with maximum frequency and probability
+            max_language = max(detected_languages, key=key_func)
+            max_probability = sum(all_language_probabilities[max_language]) / len(
+                all_language_probabilities[max_language]
+            )
+
+            # Do additional checks for silence for non-confident case
+            # calculate RMS amplitude and DC offset
+            dc_offset = audio.mean()
+            audio_minus_dc_offset = audio - dc_offset
+            is_silent = (
+                torch.all(audio.abs() < 0.01)
+                or torch.sqrt(torch.mean(audio_minus_dc_offset**2)) < 0.01
+            )
+
+            if is_silent:
+                return {"language_code": "silence", "language_confidence": 1.0}
+
+            if max_language is not None:
+                return {
+                    "language_code": max_language,
+                    "language_confidence": max_probability,
+                }
+
+        # Language is not detected for any segment and none of prev conditions met
+        return {"language_code": "silence", "language_confidence": 1.0}
```

Comment:
I guess this is better so it can evaluate to a false boolean
```suggestion
        return {"language_code": None, "language_confidence": 1.0}
```